src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py

```diff
@@ -0,0 +1,117 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""A formula pool for helping with tracking running formula engines."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Type
+
+from frequenz.channels import Sender
+
+from ...actor import ChannelRegistry, ComponentMetricRequest
+from ...microgrid.component import ComponentMetricId
+from ._formula_generators._formula_generator import (
+    FormulaGenerator,
+    FormulaGeneratorConfig,
+)
+from ._resampled_formula_builder import ResampledFormulaBuilder
+
+if TYPE_CHECKING:
+    # Break circular import by enclosing these type hints in a `TYPE_CHECKING` block.
+    from .._formula_engine import (
+        FormulaReceiver,
+        FormulaReceiver3Phase,
+        _GenericEngine,
+        _GenericFormulaReceiver,
+    )
+
+
+class FormulaEnginePool:
+    """Creates and owns formula engines from string formulas, or formula generators.
+
+    If an engine already exist with a given name, it is reused instead.
+    """
+
+    def __init__(
+        self,
+        namespace: str,
+        channel_registry: ChannelRegistry,
+        resampler_subscription_sender: Sender[ComponentMetricRequest],
+    ) -> None:
+        """Create a new instance.
+
+        Args:
+            namespace: namespace to use with the data pipeline.
+            channel_registry: A channel registry instance shared with the resampling
+                actor.
+            resampler_subscription_sender: A sender for sending metric requests to the
+                resampling actor.
+        """
+        self._namespace = namespace
+        self._channel_registry = channel_registry
+        self._resampler_subscription_sender = resampler_subscription_sender
+        self._engines: dict[str, "FormulaReceiver|FormulaReceiver3Phase"] = {}
+
+    async def from_string(
+        self,
+        formula: str,
+        component_metric_id: ComponentMetricId,
+        nones_are_zeros: bool = False,
+    ) -> "FormulaReceiver":
+        """Get a receiver for a manual formula.
+
+        Args:
+            formula: formula to execute.
+            component_metric_id: The metric ID to use when fetching receivers from the
+                resampling actor.
+            nones_are_zeros: Whether to treat None values from the stream as 0s.  If
+                False, the returned value will be a None.
```

Comment:
I wonder why docstrings does not catch double-spaces:

```suggestion
            nones_are_zeros: Whether to treat None values from the stream as 0s. If
                False, the returned value will be a None.
```

Reply:
This was not a mistake.  I went to a typing school to learn touch typing on a type writer,  so I double space my sentences. :-)

Also,  emacs, vim, etc. use double spaces to differentiate between sentence breaks and abbreviations (like `etc.` or `eg.`), when doing sentence level operations, so I like to use them too.

Reply:
alright, then let's keep it :)